src/opm/output/eclipse/WriteRFT.cpp

```diff
@@ -256,26 +344,621 @@ namespace {
+
+        bool empty(const int binId) const
+        {
+            this->verifyValid(binId);
+
+            return this->start(binId) == this->start(binId + 1);
+        }
+
+        std::optional<int> last(const int binId) const
+        {
+            this->verifyValid(binId);
+
+            if (this->empty(binId)) {
+                return {};
+            }
+
+            auto end = this->start(binId + 1);
+            return *--end;
+        }
+
+    private:
+        int minId_{std::numeric_limits<int>::max()};
+        int maxId_{std::numeric_limits<int>::min()};
+
+        std::vector<std::vector<int>::size_type> pos_{};
+        std::vector<int> ix_{};
+
+        std::vector<int>::const_iterator start(const int binId) const
+        {
+            return this->ix_.begin() + this->pos_[binId - this->minId_];
+        }
+
+        bool valid(const int binId) const
+        {
+            return (binId >= this->minId_)
+                && (binId <= this->maxId_);
+        }
+
+        void verifyValid(const int binId) const
+        {
+            if (this->valid(binId)) {
+                return;
+            }
+
+            throw std::invalid_argument {
+                fmt::format("Bin ID {} outside valid range {}..{}",
+                            binId, this->minId_, this->maxId_)
+            };
+        }
+    };
+
+    template <typename Cmp>
+    void CSRIndexRelation::build(const std::size_t              size,
+                                 const int                      minId,
+                                 const std::function<int(int)>& binId,
+                                 Cmp&&                          cmp)
+    {
+        if (size == std::size_t{0}) {
+            return;
+        }
+
+        this->ix_.resize(size);
+        std::iota(this->ix_.begin(), this->ix_.end(), 0);
+        std::sort(this->ix_.begin(), this->ix_.end(), std::forward<Cmp>(cmp));
+
+        // Sort must respect binId(i1) <= binId(i2)
+        auto inconsistentId =
+            std::adjacent_find(this->ix_.begin(),
+                               this->ix_.end(),
+                [&binId](const int i1, const int i2)
+            {
+                return binId(i1) > binId(i2);
+            });
+
+        if (inconsistentId != this->ix_.end()) {
+            throw std::invalid_argument {
+                "Comparison operator does not honour bin consistency requirement"
+            };
+        }
+
+        const auto binIdBounds =
+            std::minmax_element(this->ix_.begin(), this->ix_.end(),
+                                [&binId](const int i1, const int i2)
+                                { return binId(i1) < binId(i2); });
+
+        if (binIdBounds.first != this->ix_.end()) {
+            this->minId_ = std::min(binId(*binIdBounds.first), minId);
+        }
+
+        if (binIdBounds.second != this->ix_.end()) {
+            this->maxId_ = binId(*binIdBounds.second);
+        }
+
+        if (this->maxId_ < this->minId_) {
```

Comment:
This needs a clarifying comment. Reviewing this takes a lot of time, because I have to jump around a lot to see why the code is as complicated as it is.

I really fail to see why this can happen if we sort like above. Should this be and assert or a throw?

Reply:
> I really fail to see why this can happen if we sort like above

First, we're not sorting by `binId()` but by the `cmp` criterion.  Secondly, the `binId()` external function could, and often does, return a constant value for all elements meaning
```C++
binId(*binIdBounds.first) == binId(*binIdBounds.second)
```
This is, for instance, guaranteed to happen for multi-segmented wells which have a single branch.  If then, in addition, `binId(*binIdBounds.first) < minId` we'll end up having `maxId_ < minId_`.  This second situation is not particularly likely, but it is nevertheless possible.  I've added a comment to this effect.

> Should this be an assert or a throw?

I've switched to throwing an exception in this case.

Reply:
Second update: I've pushed a slightly different criterion here.